src/wasm32/ffi.c

```diff
@@ -282,8 +297,31 @@ ffi_call, (ffi_cif * cif, ffi_fp fn, void *rvalue, void **avalue),
     }
   }
 
-  stackRestore(structs_addr);
+  var varargs = [];
+  var varargs_addr = structs_addr;
+  for (var i = nfixedargs; i < nargs; i++) {
+    var arg_unboxed = unbox_small_structs(DEREF_U32(arg_types, i));
+    var arg_type = arg_unboxed[0];
+    var item = ffi_struct_size_and_alignment(arg_type);
+    var item_size = item[0];
+    varargs_addr -= item_size;
+    varargs.push(varargs_addr);
+    var arg_ptr = DEREF_U32(avalue, i);
+    HEAP8.subarray(varargs_addr, varargs_addr + item_size)
+        .set(HEAP8.subarray(arg_ptr, arg_ptr + item_size));
+  }
+
+  if (nfixedargs != nargs) {
+    args = args.concat(varargs);
+  }
+
+  stackRestore(varargs_addr);
+  console.log("calling fn", fn, args);
+#if WASM_BIGINT
+  var result = wasmTable.get(fn).apply(null, args);
+#else
   var result = dynCall(sig, fn, args);
+#endif
```

Comment:
The reason I unconditionally used `dynCall` is that Pyodide has this "emulate function pointer casts" patch so the incoming function pointer has a different ABI. We make a lookup table from the modified function pointers to the original ones and dereference that lookup table to get the right pointer.
https://github.com/pyodide/pyodide/blob/653891b045c46664c7e57fb084ecdade3c8b612e/emsdk/patches/fix_emscripten_dylink.patch#L11
With that patch it is necessary to use `dynCall` even with `WASM_BIGINT`.

Reply:
Ah I see, I didn't know about that patch. I reverted the `dynCall` usage since I couldn't use the linker flag `-s DEFAULT_LIBRARY_FUNCS_TO_INCLUDE='$dynCall'` within DejaGnu. It seems there is no way to escape that dollar sign (tried `$$` and `\$`, and various other combinations).

BTW, I've removed the `-s DYNCALLS` linker flag when `-s WASM_BIGINT` is also specified within commit https://github.com/hoodmane/libffi-emscripten/commit/af2325b8e2771badbf1b7609476d9ec48138107e as I think these flags are are mutually exclusive. See for example:
https://github.com/emscripten-core/emscripten/blob/660dcc00598495f5cf353771995d77a13735c0c9/src/library.js#L3503-L3507
It always calls the legacy implementation when `-s DYNCALLS` is specified, so it doesn't operate on the `wasmTable`.

Reply:
Yeah I'm actually not sure why I was having the trouble because I don't believe we are compiling with `-s WASM_BIGINT`. When we switch to `WASM_BIGINT` we'll probably just have to patch these lines in the libffi port unless we can manage to remove the emulate fpcasts patch (there's a lot of function pointer casts in the Python upstream though).

Reply:
I haven't looked at the details of that patch, but for my projects I would avoid `EMULATE_FUNCTION_POINTER_CASTS` because it can add significant runtime overhead. Nevertheless, the advantage of that linker flag is that you can compile a code base to WebAssembly faster, without having to patch upstream code.

> there's a lot of function pointer casts in the Python upstream though

fwiw, I had a similar issues with GLib and libvips. Both have been fixed with major patches on upstream code, see for example:
https://github.com/kleisauke/wasm-vips/blob/master/build/patches/glib-function-pointers.patch
https://github.com/libvips/libvips/commit/92460940337d14caf45c902ab47b8863ca41a9ae